bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Display.java

```diff
@@ -2349,6 +2361,19 @@ public static Display getDefault () {
 	}
 }
 
+/**
+ * Returns an executor that will execute any submitted runnable in the user-interface thread, regardless of the calling thread
+ * @return an {@link Executor} that executes in the user-interface thread.
+ * @since 3.120
+ */
+public static Executor executor() {
+	Display current = Display.getCurrent();
+	if (current != null) {
+		return current;
+	}
+	return Display.getDefault();
```

Comment:
I do not like that "Device implements Executor" you could as well return any other class here

Reply:
Sure but this would complicate the code/class a lot here so what specifically you don't like in Display implement executor? For sure it could be a inner class as well...

Reply:
Theoretically there could be multiple displays. One could write
`Display.getDefault().execute( something)`
but i doubt if it matters that i selected a specific display to run something.

Reply:
> Theoretically there could be multiple displays
Not only theoretical.

> Display.getDefault().execute( something)

Sure but why? make it more complicated and less flexible?

> but i doubt if it matters that i selected a specific display to run something.
At least the SWT API suggest that it could make a difference, if each display has its own thread with two displays I can run two Ui actions in parallel without blocking.

Reply:
Did you come to an agreement here guys? Even in the case of multiple displays the UI thread would be single one so not much difference.

Reply:
I'm looking at this interesting change for the 1st time and really like that
1. it provides a standard way to run in UI Thread without asking clients to verify whether they're in UI Thread or not first, so it replaces a lot of annoying code
2. It uses futures for which we can chain things very efficiently, ultimately removing need for complex chains/listeners of system vs UI jobs for patterns like "User clicks a button, a long computation happens out of the UI Thread, but we don't need a Job for it because it's in a modal dialog and we report progress differently, then the UI reacts upon completion". We happen to have it for example in Corrosion whether the wizard allows to provision the language-server.

However, I'm not sure this `executor()` static method is really useful for this story. Couldn't we just remove it if there is no clear agreement what it should do, and it there is no story to make it shiny?
I see it as yet-another-static-entry-point while we all know that it's better to infer the target from the context. Getting the display in a context (from a dialog, a widget, from an Event) is really easy, so this method wouldn't add much value, and would on the contrary lead people to write not-so-good code.
Also, I'm not knowledgeable with the multiple displays story, but I assume that in such case, it's really important that people do manage their display properly and that such method will make that code that relies on it (just as much as `Display.getDefault()` will hardly work for them). So if we can get rid of it, it would be better. `Display.getDefault()` is already "bad enough", no need to start wrapping it even more.

Reply:
I think a static method is not *strictly* required of course but:

- I frequently see `Display.getDefault()` (more than 600 references in platform) so it seems it is not *that* obvious where to get the current context display and may of those call `Display.getDefault().asyncExec` what seems to imply that one is not in an UI context yet or could acquire one other than calling a static method.
- The stream / future API makes heavy use of such static methods even if you can get it otherwise, to improve readability
- This might provides us with some opportunity here to interfere calls later on and hide some details.

Reply:
You're more motivating me to audit why some many Display.getDefault() in Platform than convincing me we need some static stuff on Display object.
I see a difference that most APIs in futures are written as utility or services and are hiding the implicit executor. But here the Display is more an object, and there can easily be multiple ones. If display implements `Executor` already, one can write
`future.thenAcceptAsync(widget::setText, widget.getDisplay())`, or `future.thenAcceptAsync(res -> findWidget().setText(res), Display.getDefault())`. I don't think introducing an API method makes it simpler in practice, and the "might provides us with some opportunity" is a bit too theoric to be a compelling reason to add one API. And even if we add an API, why not a `Display.currentOrDefault()` or why not calling it `Display.currentOrDefaultExecutor()`... This brings some more questions that we could easily just skip if we don't introduce a new API.

1 more question: what does happen if I chain `apply` on the same executors, ie if I run `CompletableFuture.completedFuture("hello").thenAcceptAsync(text -> shell::setText, shell.getDisplay()).thenRun(() -> shell.setVisible(true))`? Would that work in both UI and non-UI Thread?

Reply:
> You're more motivating me to audit why some many Display.getDefault() in Platform than convincing me we need some static stuff on Display object.

What ever fits your plans :-D

I think most uses are because code is called and caller do not know for sure if UI thread or not, something that this change should address. Maybe its just more convenient... 

> And even if we add an API, why not a `Display.currentOrDefault()` or why not calling it `Display.currentOrDefaultExecutor()`

At least I don't mind much how it is called, the rationale was just:

- If I normally would do something like `Display.getDefault()` then I could use `Display.executor()` now
- If I would call `display.anyThing` I could just pass the display

I more choose the name `executor()` as whether I get the default or the current is not of much interest when calling that method.

> what does happen if I chain `apply` on the same executors, ie if I run `CompletableFuture.completedFuture("hello").thenAcceptAsync(text -> shell::setText, shell.getDisplay()).thenRun(() -> shell.setVisible(true))`? Would that work in both UI and non-UI Thread?

I think this might describe it better than I can: https://stackoverflow.com/questions/65685715/will-a-chain-of-method-calls-completablefuture-api-execute-asynchronously-if-t